Allow skipping message body check

marcospassos opened the pull request:
Some organizations prefer to use only a subject
line, usually referring to the pull request, as
it provides the full context for the change.

Other reasons for using a subject line-only
include incompatibility with automatic PR bots,
like Renovate. Besides this, most teams today use
some UI to read the PR's message body
(i.e., GitHub), so restricting what can be added
to a PR message can lead to a poor
user experience.

Signed-off-by: Marcos Passos <marcos@croct.com>

mristin commented:
Hi @marcospassos ,
Thanks a lot for your contribution! Before I review the PR in more detail, could you quickly explain me how this differs from the already existing option `allow-one-liners: 'true'`?

mristin commented:
(The `allow-one-liners` is documented here in the readme: https://github.com/mristin/opinionated-commit-message#one-liners)

Fryuni commented:
The `allow-one-liners` still checks the body if it exists, only skipping the validation if it is empty.

One use case for this is squash merge of PRs, where GH will keep only the title of the PR and not the body. The body of the PR might be invalid as a commit body, but it won't be used as such (long lines for tables, links and images for example).

I work with Marcos, this is related to #96 that I opened. Two different use cases that we have depending on the project

mristin commented on README.md:
````diff
@@ -158,6 +158,21 @@ You can allow one-liner commit messages by setting the flag `allow-one-liners`:
           allow-one-liners: 'true'
 ```
 
+## Skip body check
````
```suggestion
## Skip Body Check
```

mristin commented on README.md:
```diff
@@ -158,6 +158,21 @@ You can allow one-liner commit messages by setting the flag `allow-one-liners`:
 
+## Skip body check
+
+Some organizations prefer to use only a subject line, usually referring to the pull request, 
+as it provides the full context for the change.
```
```suggestion
For some repositories only the subject matters while the body is allowed to be free-form.
For example, this is the case when the body of the commit is automatically generated (*e.g.*, by a third-party service that we do not control).
In such situations, we want check the subject line, but ignore the body in the checks.
```

mristin commented on src/__tests__/inspection.test.ts:
```diff
@@ -55,14 +55,28 @@ it('reports too few lines with disallowed one-liners.', () => {
+  const message =
+    'Change SomeClass to OtherClass\n' +
+    '\n' +
+    'This replaces the SomeClass with OtherClass in all of the module ' +
+    'since Some class was deprecated. This is long message should not ' +
+    'be checked.';
```
```suggestion
    'Change SomeClass to OtherClass.';
```

This would have triggered an error without the `skip-body-check` -- could you please double-check?

mristin commented on src/__tests__/inspection.test.ts:
```diff
@@ -55,14 +55,28 @@ it('reports too few lines with disallowed one-liners.', () => {
 });
 
+it('reports no errors on any message when body check is disabled.', () => {
+  const inputs = input.parseInputs('', '', 'true', '', 'true').mustInputs();
```
```suggestion
  const inputs = input.parseInputs('', '', '', '', 'true').mustInputs();
```

If I'm not mistaken, you don't have to explicitly set `allowOneLinersInput`?

mristin commented on src/__tests__/main.test.ts:
```diff
@@ -69,6 +69,25 @@ it('considers allow-one-liners.', () => {
+  (commitMessages.retrieve as any).mockImplementation(() => [
+    'Change SomeClass to OtherClass\n' +
+      '\n' +
+      'This replaces the SomeClass with OtherClass in all of the module ' +
+      'since Some class was deprecated.'
```
```suggestion
      'Change SomeClass to OtherClass'
```

The repetition of the subject should trigger an error unless we set the `skip-body-check`.

mristin reviewed (changes requested):
@marcospassos thanks again for your contribution! I noticed a couple of minor mistakes in the tests and made the corresponding suggestions.

marcospassos commented:
I believe all issues have been addressed.

mristin reviewed (approved):
LGTM!

mristin commented:
@marcospassos sorry, I've just noticed that powershell pendant has not been updated. Let me think over the weekend whether I should remove it altogether. The local powershell script is not really that useful, but is quite a maintenance burden.

marcospassos commented:
We don't use it either :/

mristin commented:
@marcospassos I discontinued the local scripts in #98. Could you please rebase on master and re-run the tests to make sure that everything works correctly?

mristin commented on src/__tests__/main.test.ts:
```diff
@@ -69,6 +69,26 @@ it('considers allow-one-liners.', () => {
+  const mockSetFailed = jest.fn();
+  (core as any).setFailed = mockSetFailed;
+
+  (core as any).getInput = (name: string) =>
```
(Please see the unit test that fails for the context.) This seems to be set for all the subsequent unit tests and triggers the unit tests below to fail. 

I'm not that experienced with Jest -- do you know what would be an idiomatic way to mock `core` such that individual unit tests can be run in parallel and don't affect each other?

A quick search lead me to this StackOverflow question: https://stackoverflow.com/questions/67260803/is-it-possibly-for-jest-mocks-to-be-overwritten-by-each-other

It seems to me that the right approach would be to use `mockImplementationOnce`. Please give me your feedback first and then I'll re-write the `main.test.ts` first before we make changes in this pull request.

mristin replied:
Actually, on a second read, I should prefix the test with `beforeEach` and reset the mocks. I'll try to do that today and come back to you once I fixed the tests.

marcospassos commented:
As you've pointed out, the mock states were leaking to subsequent tests. 

One of the main reasons was that you are not using the most idiomatic mocking in Jest. By assigning a mock, Jest can't track down what is mocked and where, so there is no way to restore/reset the mocked methods afterward. Note that you don't need to mock the modules on top. That's why I removed those lines.

I've also refactored the asserts to prove richer errors messages on failures.

Hopefully, all tests should be passing now.

mristin commented:
@marcospassos thanks! Would you like to make a separate PR for the refactoring or you'd prefer that I create it and put you as co-author?

mristin commented:
@marcospassos thanks again for the help with the idiomatic tests. I merged them in the PR #100. Could you please rebase just to make sure that nothing breaks?

marcospassos commented:
I think we're ready to go.

mristin commented:
Thanks again, @marcospassos !

marcospassos commented:
@mristin, is there any chance of releasing a new version so we can stop using our fork?

mristin commented:
@marcospassos sure, let us merge in #96 and then I'll release a new major version (major since local scripts were discontinued).
